cli/internal/fs/turbo_json.go

```diff
@@ -57,8 +77,30 @@ type rawTask struct {
 	Persistent bool                `json:"persistent"`
 }
 
+// rawTask exists to Unmarshal from json. When fields are omitted, we _want_
+// them to be missing, so that we can distinguish missing from empty value.
+type rawTask struct {
+	Outputs    []string             `json:"outputs,omitempty"`
+	Cache      *bool                `json:"cache,omitempty"`
+	DependsOn  []string             `json:"dependsOn,omitempty"`
+	Inputs     []string             `json:"inputs,omitempty"`
+	OutputMode *util.TaskOutputMode `json:"outputMode,omitempty"`
+	Env        []string             `json:"env,omitempty"`
+	Persistent bool                 `json:"persistent,omitempty"`
+}
+
+// PristinePipeline contains original TaskDefinitions without the bookkeeping
+type PristinePipeline map[string]TaskDefinition
+
 // Pipeline is a struct for deserializing .pipeline in configFile
-type Pipeline map[string]TaskDefinition
+type Pipeline map[string]BookkeepingTaskDefinition
+
+// BookkeepingTaskDefinition holds the underlying TaskDefinition and some bookkeeping data
+// about the TaskDefinition. This wrapper struct allows us to leave TaskDefinition untouched.
+type BookkeepingTaskDefinition struct {
+	fieldsMeta     map[string]bool
```

Comment:
It looks like we're just using this as a set and never care about the value. There's a [slight win from using `map[string]struct{}`](https://itnext.io/set-in-go-map-bool-and-map-struct-performance-comparison-5315b4b107b) instead, but the ergonomics of it are a little weird. If that change would be invasive, I'd be fine with just a comment clarifying that this map is used as a set and the values don't matter.

Reply:
We also have a `Set` library

Reply:
updated to use `util.Set`